### PR TITLE
WIP: interchange: add BELs chain support

### PR DESCRIPTION
**Pull request by kowalewskijan**
Initial support for BELs chains.
Related to:
* https://github.com/SymbiFlow/nextpnr/issues/262
* https://github.com/SymbiFlow/python-fpga-interchange/pull/70

Signed-off-by: Jan Kowalewski <jkowalewski@antmicro.com>

### CODE REVIEW

File: interchange/DeviceResources.capnp [hunk 1]
```diff
@@ -73,22 +73,23 @@ using TileTypeSiteTypeIdx = UInt32;
...
+  nodes                 @6 : List(Node);
+  primLibs              @7 : Dir.Netlist; # Netlist libraries of Unisim primitives and macros
+  exceptionMap          @8 : List(PrimToMacroExpansion); # Prims to macros expand w/same name, except these
+  cellBelMap            @9 : List(CellBelMapping);
+  cellInversions       @10 : List(CellInversion);
+  packages             @11 : List(Package);
+  constants            @12 : Constants;
+  constraints          @13 : Constraints;
+  lutDefinitions       @14 : LutDefinitions;
+  belChainsDefinitions @15 : BelChainsDefinitions;
```
Comment (gatecat) [inline interchange/DeviceResources.capnp:91]: why is this before `parameterDefs`? afaik that will cause a backwards compatibility issue

File: interchange/DeviceResources.capnp [hunk 2]
```diff
@@ -627,6 +628,49 @@ struct Device {
...
   }
 
+  ######################################
+  # BEL chains definitions
+  ######################################
+  struct BelChainsDefinitions {
+    struct BelChain {
+
+      struct ChainPattern {
+        source @0 : Text;
```
Comment (gatecat) [inline interchange/DeviceResources.capnp:638]: the indentation here is inconsistent

File: interchange/DeviceResources.capnp [hunk 3]
```diff
@@ -627,6 +628,49 @@ struct Device {
...
+	sink   @1 : Text;
+      }
+
+      struct ChainCoordConfig {
+        coord @0 : Text;
```
Comment (gatecat) [inline interchange/DeviceResources.capnp:643]: this should be an enum, not text

File: interchange/DeviceResources.capnp [hunk 4]
```diff
@@ -627,6 +628,49 @@ struct Device {
...
+      }
+
+      struct ChainCellPort {
+        name  @0 : Text;
+	width @1 : Int8;
+      }
+
+      struct ChainCell {
+        cell        @0 : Text;
+	inputPins   @1 : List(ChainCellPort);
```
Comment (gatecat) [inline interchange/DeviceResources.capnp:654]: it should be possible to get a list of ports from `primLibs` and/or the cell->bel mappings; I don't know why these are needed unless the intent is to refer to a subset of ports only?

**Review by gatecat — COMMENTED** [review]
see comments, it would also be a good idea to have some more examples of how this is supposed to work, in particular some clarity as to how cells, bels, sites and tiles all apply here

File: interchange/DeviceResources.capnp [hunk 5]
```diff
@@ -81,23 +81,24 @@ using TileTypeSiteTypeIdx = UInt32;
...
-  exceptionMap    @8 : List(PrimToMacroExpansion); # Prims to macros expand w/same name, except these
-  cellBelMap      @9 : List(CellBelMapping);
-  cellInversions @10 : List(CellInversion);
-  packages       @11 : List(Package);
-  constants      @12 : Constants;
-  constraints    @13 : Constraints;
-  lutDefinitions @14 : LutDefinitions;
-  parameterDefs  @15 : ParameterDefinitions;
-  wireTypes      @16 : List(WireType);
+  name                  @0 : Text;
```
Comment (mithro) [inline interchange/DeviceResources.capnp:84]: Why is the indenting here changing? Can you submit the reformatting as a separate pull request?

File: interchange/DeviceResources.capnp [hunk 6]
```diff
@@ -656,6 +657,46 @@ struct Device {
     lutElements @1 : List(LutElements);
   }
 
+  ######################################
+  # BEL chains definitions
+  ######################################
```
Comment (mithro) [inline interchange/DeviceResources.capnp:662]: Do you have any information about why this is needed? This all seems easily detectable from the routing?

**Review by mithro — CHANGES_REQUESTED** [review]

File: interchange/DeviceResources.capnp (hunk 6 above)
Comment (kowalewskijan) [reply interchange/DeviceResources.capnp:662]: I’m not sure if I understand it correctly. The idea is to have a set of chain definitions like carries and any other chainable BELs in the schema to provide information to PnR tool which cells can be clusterized and placed as a cluster/chain.

AFAIK it is possible to detect chained cells by checking if e.g. cell has a driver which is another cell of the same type and iterate until we hit a constant driver. If this is what you meant then it is doable, but this approach would probably take more time to detect them rather than providing in schema which cell types are chainable, picking the right ones and clustering them together.

File: interchange/DeviceResources.capnp (hunk 5 above)
Comment (kowalewskijan) [reply interchange/DeviceResources.capnp:84]: Names, IDs, values are aligned - by adding the new field `belChainsDefinitions` which has longer name than any other existing field, I assumed that fields may need to be re-aligned to the new one.

**Comment by gatecat** [comment]
My understanding is that this has now become part of `device_config.yaml`, in SymbiFlow/python-fpga-interchange#91. While we might want to bring this back into the schema long term, I think it's best if this PR is closed for now?
